src/lib.rs

```diff
@@ -439,23 +504,102 @@ impl Lockfile {
     }
   }
 
-  pub fn new(opts: NewLockfileOptions) -> Result<Lockfile, Box<LockfileError>> {
-    fn load_content(
+  pub async fn new(
+    opts: NewLockfileOptions<'_>,
+    provider: &dyn NpmPackageInfoProvider,
+  ) -> Result<Lockfile, Box<LockfileError>> {
+    async fn load_content(
       content: &str,
+      provider: &dyn NpmPackageInfoProvider,
     ) -> Result<LockfileContent, LockfileErrorReason> {
       let value: serde_json::Map<String, serde_json::Value> =
         serde_json::from_str(content)
           .map_err(LockfileErrorReason::ParseError)?;
       let version = value.get("version").and_then(|v| v.as_str());
       let value = match version {
-        Some("4") => value,
-        Some("3") => transforms::transform3_to_4(value)?,
+        Some("5") => value,
+        Some("4") => transforms::transform4_to_5(value, provider).await?,
+        Some("3") => {
+          transforms::transform4_to_5(
+            transforms::transform3_to_4(value)?,
+            provider,
+          )
+          .await?
+        }
         Some("2") => {
-          transforms::transform3_to_4(transforms::transform2_to_3(value))?
+          transforms::transform4_to_5(
+            transforms::transform3_to_4(transforms::transform2_to_3(value))?,
+            provider,
+          )
+          .await?
+        }
+        None => {
+          transforms::transform4_to_5(
+            transforms::transform3_to_4(transforms::transform2_to_3(
+              transforms::transform1_to_2(value),
+            ))?,
+            provider,
+          )
+          .await?
+        }
+        Some(version) => {
+          return Err(LockfileErrorReason::UnsupportedVersion {
+            version: version.to_string(),
+          });
+        }
+      };
+      let content = LockfileContent::from_json(value.into())
+        .map_err(LockfileErrorReason::DeserializationError)?;
+
+      Ok(content)
+    }
+
+    // Writing a lock file always uses the new format.
+    if opts.overwrite {
+      return Ok(Lockfile {
+        overwrite: opts.overwrite,
+        filename: opts.file_path,
+        has_content_changed: false,
+        content: LockfileContent::default(),
+      });
+    }
+
+    if opts.content.trim().is_empty() {
+      return Err(Box::new(LockfileError {
+        file_path: opts.file_path.display().to_string(),
+        source: LockfileErrorReason::Empty,
+      }));
+    }
+    let content =
+      load_content(opts.content, provider)
+        .await
+        .map_err(|reason| LockfileError {
+          file_path: opts.file_path.display().to_string(),
+          source: reason,
+        })?;
+    Ok(Lockfile {
+      overwrite: opts.overwrite,
+      has_content_changed: false,
+      content,
+      filename: opts.file_path,
+    })
+  }
+
+  pub fn new_current_version(
```

Comment:
added this to have a sync way to make a lockfile. if it needs to do the async transform, it returns an error (which the caller needs to handle). this is also useful for tests and other places where you know the lockfile is already on the latest version